src/encord_active/analysis/executor.py

```diff
@@ -0,0 +1,1177 @@
+            hash_collision=hash_collision,
+        )
+
+    def _stage_2(self):
+        # Stage 2: Support for complex value based dependencies.
+        # Note: in-efficient, but needed for first version of this.
+        # Stage 2 evaluation will always be very closely bound to the executor
+        # and the associated storage and indexing strategy.
+        for metric in tqdm(self.config.derived_embeddings, desc="Metric Computation Stage 2: Building indices"):
+            metrics_tqdm_desc = f"                          : Generating indices for {metric.ident}"
+            metrics_tqdm = tqdm([self.data_metrics, self.annotation_metrics], desc=metrics_tqdm_desc)
+            for metrics in metrics_tqdm:
+                metrics_tqdm.set_description(metrics_tqdm_desc)
+                if len(metrics) == 0:
+                    continue
+                if isinstance(metric, NearestImageEmbeddingQuery):
+                    # Embedding query lookup
+                    max_neighbours = int(metric.max_neighbours)
+                    if max_neighbours > 50:
+                        raise ValueError(f"Too many embedding nn-query results: {metric.ident} => {max_neighbours}")
+                    data_keys = list(metrics.keys())
+                    metrics_tqdm.set_description(f"{metrics_tqdm_desc}  [Loading data]")
+                    data_list = [metrics[k][metric.embedding_source].cpu().numpy() for k in data_keys]
+                    metrics_tqdm.set_description(f"{metrics_tqdm_desc}  [Preparing data]")
+                    data = np.stack(data_list)
+                    # FIXME: tune threshold where using pynndescent is actually faster.
+                    if len(data_list) <= 4096:
+                        # Exhaustive search is faster & more correct for very small datasets.
+                        metrics_tqdm.set_description(f"{metrics_tqdm_desc}  [Exhaustive search]")
+                        data_indices_list = []
+                        data_distances_list = []
+                        for data_key, data_embedding in zip(data_keys, data_list):
+                            data_raw_offsets = data - data_embedding
+                            data_raw_similarities = np.linalg.norm(data_raw_offsets, axis=1)
+                            data_raw_index = np.argsort(data_raw_similarities)[:max_neighbours]
+                            data_indices_list.append(data_raw_index)
+                            data_distances_list.append(data_raw_similarities[data_raw_index])
+                        data_indices = np.stack(data_indices_list)
+                        data_distances = np.stack(data_distances_list)
+                    else:
+                        metrics_tqdm.set_description(f"{metrics_tqdm_desc}  [Creating index]")
+                        nn_descent = NNDescent(
+                            data=data,
+                            parallel_batch_queries=True,
+                            n_jobs=-1,
+                        )
```

Comment:
Shall we set metric='cosine', so that it is bounded in [-1, 1]?

Reply:
added option for different distance metrics